src/main/java/macrobase/analysis/stats/TreeKDE.java

```diff
@@ -79,9 +81,21 @@ public KDTree getKdtree() {
     }
 
     @Override
+    /**
+     * Return the negative log pdf density, this avoids underflow errors while still being
+     * an interpretable quantity. Use scoreDensity if you need the actual negative pdf.
+     */
     public double score(Datum datum) {
+        numScored++;
+        if (numScored % 10000 == 0) {
+            log.debug("Scored {}", numScored);
+        }
         double unscaledScore = scoreKDTree(kdtree, datum);
         // Note: return score with a minus sign, s.t. outliers are selected not inliers.
-        return -unscaledScore * scoreScalingFactor;
+        return -(Math.log(unscaledScore) + scoreScaleLog);
+    }
+
+    public double scoreDensity(Datum datum) {
+        return -Math.exp(-score(datum));
```

Comment:
So that it's still a "score" and outliers have high scores (and to make it consistent with the score KDE returns). I'm refactoring the KDE and TreeKDE class now and I can change the function to just density(Datum datum) though and just return the actual true density since we don't really need a negative density anywhere.